lib/jsdom/living/helpers/http-request.js

```diff
@@ -0,0 +1,261 @@
+  constructor(url, options) {
+    super();
+    this._currentURL = url;
+    this._options = options;
+    this.headers = this._options.headers;
+    this._ended = false;
+    this._redirectCount = 0;
+    this._requestBodyBuffers = [];
+    this._bufferIndex = 0;
+    this.response = Object.create(null);
+    this._performRequest();
+  }
+
+  abort() {
+    abortRequest(this._currentRequest);
+    this.emit("abort");
+    this.removeAllListeners();
+  }
+
+  write(data, encoding) {
+    if (data.length > 0) {
+      this._requestBodyBuffers.push({ data, encoding });
+      this._currentRequest.write(data, encoding);
+    }
+  }
+
+  end() {
+    this.emit("request", this._currentRequest);
+    this._ended = true;
+    this._currentRequest.end();
+  }
+
+  setHeader(name, value) {
+    this._options.headers[name] = value;
+    this._currentRequest.setHeader(name, value);
+  }
+
+  removeHeader(name) {
+    delete this._options.headers[name];
+    this._currentRequest.removeHeader(name);
+  }
+
+  // Without this method, the test suite will halt
+  toJSON() {
+    const { method, headers } = this._options;
+    return { uri: new URL(this._currentURL), method, headers };
+  }
+
+  _writeNext(error) {
+    if (this._currentRequest) {
+      if (error) {
+        this.emit("error", error);
+      } else if (this._bufferIndex < this._requestBodyBuffers.length) {
+        const buffer = this._requestBodyBuffers[this._bufferIndex++];
+        if (!this._currentRequest.writableEnded) {
+          this._currentRequest.write(
+            buffer.data,
+            buffer.encoding,
+            this._writeNext.bind(this)
+          );
+        }
+      } else if (this._ended) {
+        this._currentRequest.end();
+      }
+    }
+  }
+
+  _performRequest() {
+    const urlOptions = new URL(this._currentURL);
+    const scheme = urlOptions.protocol;
+    this._options.agent = this._options.agents[scheme.substring(0, scheme.length - 1)];
+    const { request } = scheme === "https:" ? https : http;
+    this._currentRequest = request(this._currentURL, this._options, response => {
+      this._processResponse(response);
+    });
+
+    let cookies;
+    if (this._redirectCount === 0) {
+      this.originalCookieHeader = this.getHeader("Cookie");
+    }
+    const jar = this._options.cookieJar;
+    if (jar) {
+      cookies = jar.getCookieStringSync(this._currentURL);
+    }
+    if (cookies && cookies.length) {
+      if (this.originalCookieHeader) {
+        this.setHeader("Cookie", this.originalCookieHeader + "; " + cookies);
+      } else {
+        this.setHeader("Cookie", cookies);
+      }
+    }
+
+    for (const event of ["connect", "error", "socket", "timeout"]) {
+      this._currentRequest.on(event, (...args) => {
+        this.emit(event, ...args);
+      });
+    }
+    if (this._isRedirect) {
+      this._bufferIndex = 0;
+      this._writeNext();
+    }
+  }
+
+  _processResponse(response) {
+    response.request = {};
```

Comment:
Why is this stored as a property of `response`? It seems like it's only used in this function and so could be a normal variable?

Reply:
Required by XMLHTTPRequestImpl.js, but nowhere else (I think). There might yet be a way to refactor the XHR module such that this parameter patching can be circumvented.